### PR TITLE
Randomization of parameters

### PR DESCRIPTION
**Pull request by Bonnevie**
As per issue #367, this pull request implements randomization functionality for parameters. The code is minimal, with parameters now being able to randomly reset themselves and models being able to randomize all children. Priors have been modified to have a sample method.   

If possible it samples from priors, otherwise it defaults to a weak but universal alternative where it samples a free unconstrained state from a standard normal and then propagates it through the forward transform to obey constraints.  

The implemented tests verify that parameters are modified, that it obeys "fixed" flags, that priors supersede default behavior, that all priors work, and that randomization can be passed down through the model hierarchy appropriately.

edit: codeship appears to be throwing an error, but I cannot view the details of it without admin privileges?

### CODE REVIEW

**Comment by jameshensman** [comment]
This doesn't appear to work for params that have size greater than 1. 


```python
In [1]: m = GPflow.param.Parameterized()

In [2]: m.p = GPflow.param.Param(np.random.randn(3))

In [3]: print(m)
unnamed.p transform:(none) prior:None
[ 0.40870328 -0.25011942  0.03627796]

In [4]: m.randomize()

In [5]: print(m)
unnamed.p transform:(none) prior:None
0.7736654332138937
```

**Comment by Bonnevie** [comment]
That's a pretty silly oversight on my part... 
What's the assumption for priors on multivariate quantities? i.i.d.?

**Comment by jameshensman** [comment]
yes.

**Comment by Bonnevie** [comment]
@jameshensman: got it fixed now.

File: GPflow/param.py [hunk 1]
```diff
@@ -284,6 +284,30 @@ def set_state(self, x):
...
+                Defaults to True.
+        """
+        if not (skipfixed and self.fixed):
+            if self in distributions.keys():
+                self._array = distributions[self].sample(self.shape)
+            else:
+                try:
+                    self._array = self.prior.sample(self.shape)
+                except AttributeError:
+                    randn = np.array(np.random.randn(*self.shape))
```
Comment (jameshensman) [inline GPflow/param.py]: Some transforms alter the size of their argument. I think you need `self.transform.free_state_size(self.shape)` to get the correct number of randns here.

File: GPflow/priors.py [hunk 2]
```diff
@@ -75,6 +88,9 @@ def __init__(self, shape, scale):
     def logp(self, x):
         return tf.reduce_sum(densities.gamma(self.shape, self.scale, x))
 
+    def sample(self, shape=()):
+        return np.array(np.random.gamma(self.shape, self.scale))
```
Comment (jameshensman) [inline GPflow/priors.py]: shape arg is not used.

File: GPflow/priors.py [hunk 3]
```diff
@@ -88,6 +104,10 @@ def __init__(self, mu, sigma):
     def logp(self, x):
         return tf.reduce_sum(densities.laplace(self.mu, self.sigma, x))
 
+    def sample(self, shape=()):
+        return np.array(np.random.laplace(self.mu, self.sigma,
+                                          size=shape if shape else None))
```
Comment (jameshensman) [inline GPflow/priors.py]: Why wrap this in np.array? wouldn't it be neater to have the shape default to (1,) ?

File: testing/test_param.py [hunk 4]
```diff
@@ -599,6 +599,107 @@ def test(self):
...
+        m.p2 = GPflow.param.Param(1.0)
+        m.m = GPflow.model.Model()
+        m.m.p = GPflow.param.Param(1.0)
+        m.m.p2 = GPflow.param.Param(1.0)
+
+        m.p2.prior = GPflow.priors.Gaussian(0,1)
+        m.m.p2.prior = GPflow.priors.Gaussian(0,1)
+        m.randomize()
+
+        self.assertTrue(m.p.value != 1.0)
```
Comment (jameshensman) [inline testing/test_param.py]: I think `self.assertFalse(m.p.value == 1.0)` is cleaner :)

**Comment by jameshensman** [comment]
Hi @Bonnevie , 

This is pretty good!  I like that you've used the style used in the rest of GPflow, e.g. the abstract base classes, and you've made good use of the tree structure.  Thanks for adding some pretty comprehensive tests too. 

I added a few comments to address some specific things. Please also add a test that covers the shaping that I pointed out in the comment above (i.e. a test that would have failed in the previous version of the code).

**Comment by Bonnevie** [comment]
@jameshensman: sorry for the weekend silence. I did add a few new lines to the tests to take multi-dimensional parameters into account, but I can see from review that I forgot an assert. Will look at your other comments :)

File: GPflow/priors.py (hunk 3 above)
Comment (Bonnevie) [inline GPflow/priors.py]: I believe (1,) gives you a 1-d vector with one element which is ever so slightly different from a numpy scalar (e.g. you can index the latter).
I wanted scalars to be default, but I also wanted everything to be Numpy for consistency.

Comment (Bonnevie) [inline GPflow/priors.py]: Checked what the default is for scalar Parameters in GPflow, and that's actually (1,) so I'll modify it to fit that standard.

**Comment by Bonnevie** [comment]
@jameshensman: I think I covered all of your comments. Dropped the np.array wrapping and changed the default shape.

File: GPflow/param.py [hunk 5]
```diff
@@ -305,7 +305,8 @@ def randomize(self, distributions={}, skipfixed=True):
                 try:
                     self._array = self.prior.sample(self.shape)
                 except AttributeError:
-                    randn = np.array(np.random.randn(*self.shape))
+                    randn = np.array(np.random.randn(
```
Comment (jameshensman) [inline GPflow/param.py]: I don't think we need `np.array(...)` here.

**Comment by jameshensman** [comment]
One teeny extra `array(...)` that we dont need :)

**Comment by Bonnevie** [comment]
ah, yeah, that shouldn't matter anymore either. Removed it!

**Comment by Bonnevie** [comment]
all of a sudden I'm getting some peculiar errors with respect to "referenceRbfKernel" and stuff like that.

**Comment by jameshensman** [comment]
The failing tests were a problem with tokens on coveralls, which is an alternative coverage site to codecov. Not your code :)

I think this is ready to merge.

**Comment by thj2009** [comment]
Hi, do you have an example or test to explain how to use the restarting optimization function?
Thank you very much.

**Pull request merged**